Phases/add ranged attack

45930 opened the pull request:
Prime: @louiswheeleriv 

I might not have every corner case covered yet, but I want to make fast progress on the core logic of these phases and start getting integrated with the full stack.  I also want to give you the preview on how dice rolls can be used.  The test has an example of how the rolls are _encrypted_ and the phase action has an example of how the rolls are _decrypted_.

45930 commented on src/phase/PhaseState.ts:
```diff
@@ -119,6 +129,109 @@ export class PhaseState extends Struct({
+
+    let healthDiff = Circuit.if(
+      Bool.and(Bool.and(hit, wound), notSave),
+      UInt32.from(2),
```
made up damage value.  I will consult the PDR for the real version

45930 commented on src/phase/PhaseState.ts:
```diff
@@ -119,6 +129,109 @@ export class PhaseState extends Struct({
+    // roll for save
+    const notSave = Circuit.if(
+      decrytpedRolls.save.greaterThanOrEqual(
+        attackingPiece.condition.saveRoll.value
```
Not counting armor piercing for now

45930 commented:
I split the phase state tests into each method getting its own file.  I could have done more nested describe blocks, but I thought the files were more readable.

louiswheeleriv commented on src/phase/PhaseState.ts:
```diff
@@ -119,6 +129,109 @@ export class PhaseState extends Struct({
+    );
+
+    // roll for save
+    const notSave = Circuit.if(
+      decrytpedRolls.save.greaterThanOrEqual(
+        attackingPiece.condition.saveRoll.value
+      ),
+      Bool(false),
+      Bool(true)
+    );
```
Totally okay if we're leaving out armor piercing for now, but this will need to reference the `saveRoll` of the target piece, not that of the attacker.

Example:
```
saveRollThreshold = targetPiece.saveRoll + attackingPiece.armorPiercing
saveRollActual = rollDice()
saveRollPassed = saveRollActual >= saveRollThreshold
if (!saveRollPassed) applyDamage()
```
Just pseudocode and obviously not compatible with snarkyjs but hopefully that explains the general behavior

45930 replied:
Yes good catch!

louiswheeleriv reviewed (commented):
Looking great! Tests look super helpful. One comment on the save roll, let me know if you want to chat about that.

45930 commented:
@louiswheeleriv - I fixed the bug you mentioned.  I also added the rangedDamage attribute to the unit stats and added more tests for the different behavior we want.
